clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3177,25 +3177,32 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
 
   ASTContext &Ctx = S.getASTContext();
 
-  if (!XDimExpr->isValueDependent() && !YDimExpr->isValueDependent() &&
```

Comment:
This is doing something subtly different than before. Before, if any of the expressions were value dependent, we didn't bother trying to verify any of the constant expressions. Now, we consider each dimension independently, which will change testing behavior (we can get diagnostics when we previously did not).

I'm not saying this is a bad change in behavior, but it is a difference in behavior and it's possible the CI pipeline is failing because of it.

Btw, this suggests we're missing test coverage that should be added as part of the patch. Can you add a test case that would crash without your changes so that we can verify that your fix doesn't regress some day in the future?

Reply:
I'm trying to figure out exactly how this behaviour is different..

Will add a test such that YDimExpr and ZDimExpr will be nullptr in the case where sycl:: or intel:: reqd_work_group_size spelling is not used. Sounds ok?

Reply:
> I'm trying to figure out exactly how this behaviour is different..

It differs because the previous code would not verify the integer constant expressions of `y` or `z` unless all of x, y, and z are not value dependent. In the new code, we check them individually, so if x is not value dependent, but y and z are, we'll get errors before template instantiation instead of after (which may or may not be a bad thing!).

> Will add a test such that YDimExpr and ZDimExpr will be nullptr in the case where sycl:: or intel::reqd_work_group_size spelling is not used. Sounds ok?

That sounds sensible to me.

Reply:
Right, the idea is you want to only do the checking if the three dimensions are either null or not-value-dependent, otherwise you should skip the checking.  Then, make sure you check them each individually, but only if they are non-null.

Reply:
@erichkeane Are you suggesting a reversal of my order of operations, i.e., doing the individual checking within the if(!XDimExpr->isValueDependent() && YDimExprIsDereferencable && ZDimExprIsDerferencable) ?

Reply:
there is a top level check before all the checking to make sure that everything is either null or not-dependent. So something like:

    if ((XDimExpr && XDimExpr->isValueDependent()) | |YDimExpr && YDimExpr->isValueDependent()) || ZDimExpr && ZDimExpr->isValueDependent()))
        //skip checking.


Then when inspecting each value, you only check it if non-null.

Reply:
@erichkeane 
Couldn't find a case where YDimExpr or ZDimExpr can actually be nullptr: 

1. For non sycl:: or non intel:: use, such as the OpenCL spelling __attribute__((reqd_work_group_size(X,Y,Z)), three arguments are mandatory. 
2. For sycl:: or intel:: usage, when the user doesn't pass Y or Z arguments, there is a default initialization to 1
3. The possibility of nullptr in SetDefaultValue definition seems only theoretical

Reply:
Agreed; I think the Klokwork report is a false positive, but we could eliminate it (hopefully) and improve the readability of the code by using an assert within the lambda to assert that the only way to call the lambda is for an attribute that *can* elide the Y and Z arguments. Then we can remove the return of `nullptr` from the lambda and simplify that code a bit.

Reply:
In that case, then yes, I agree with Aaron: We need to make it so that lambda doesn't have the ability to return nullptr.

Reply:
Modified the lambda such that it will not return a nullptr; added an assert.